finality-aleph/src/abft/network.rs

```diff
@@ -0,0 +1,74 @@
+use std::marker::PhantomData;
+
+use log::warn;
+use sp_runtime::traits::Block;
+
+use crate::{
+    abft::SignatureSet,
+    crypto::Signature,
+    data_io::{AlephData, AlephNetworkMessage},
+    network::{Data, DataNetwork},
+    Hasher,
+};
+
+pub type LegacyNetworkData<B> =
+    legacy_aleph_bft::NetworkData<Hasher, AlephData<B>, Signature, SignatureSet<Signature>>;
+
+pub type CurrentNetworkData<B> =
+    current_aleph_bft::NetworkData<Hasher, AlephData<B>, Signature, SignatureSet<Signature>>;
+
+impl<B: Block> AlephNetworkMessage<B>
+    for legacy_aleph_bft::NetworkData<Hasher, AlephData<B>, Signature, SignatureSet<Signature>>
+{
+    fn included_data(&self) -> Vec<AlephData<B>> {
+        self.included_data()
+    }
+}
+impl<B: Block> AlephNetworkMessage<B>
+    for current_aleph_bft::NetworkData<Hasher, AlephData<B>, Signature, SignatureSet<Signature>>
+{
+    fn included_data(&self) -> Vec<AlephData<B>> {
+        self.included_data()
+    }
+}
+
+/// A wrapper needed only because of type system theoretical constraints. Sadness.
+pub struct NetworkWrapper<D: Data, DN: DataNetwork<D>> {
+    inner: DN,
+    _phantom: PhantomData<D>,
+}
+
+impl<D: Data, DN: DataNetwork<D>> From<DN> for NetworkWrapper<D, DN> {
+    fn from(inner: DN) -> Self {
+        NetworkWrapper {
+            inner,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<D: Data, DN: DataNetwork<D>> current_aleph_bft::Network<D> for NetworkWrapper<D, DN> {
+    fn send(&self, data: D, recipient: current_aleph_bft::Recipient) {
+        if self.inner.send(data, recipient.into()).is_err() {
```

Comment:
Consider decreasing duplication by introducing helper generic functions in base struct, e.g.:

```rust
impl<D: Data, DN: DataNetwork<D>> NetworkWrapper<D, DN> {
    fn send<R>(&self, data: D, recipient: R)
    where
        crate::abft::types::Recipient: From<R>,
    {
        if self.inner.send(data, recipient.into()).is_err() {
            warn!(target: "aleph-network", "Error sending an AlephBFT message to the network.");
        }
    }

    async fn next_event(&mut self) -> Option<D> {
        self.inner.next().await
    }
}

#[async_trait::async_trait]
impl<D: Data, DN: DataNetwork<D>> current_aleph_bft::Network<D> for NetworkWrapper<D, DN> {
    fn send(&self, data: D, recipient: current_aleph_bft::Recipient) {
        self.send(data, recipient);
    }

    async fn next_event(&mut self) -> Option<D> {
        self.next_event().await
    }
}

#[async_trait::async_trait]
impl<D: Data, DN: DataNetwork<D>> legacy_aleph_bft::Network<D> for NetworkWrapper<D, DN> {
    fn send(&self, data: D, recipient: legacy_aleph_bft::Recipient) {
        self.send(data, recipient);
    }

    async fn next_event(&mut self) -> Option<D> {
        self.next_event().await
    }
}
```

Reply:
done here and in partially for keychain where it was applicable